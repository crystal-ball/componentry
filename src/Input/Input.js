import React, { createContext, useContext, useRef } from 'react'
import nanoid from 'nanoid'
import elem from '../elem-factory'
import { useTheme } from '../Theme/Theme'

const InputCtx = createContext({ guid: null })

export default function Input({ children }) {
  /**
   * Guid instance property will be uniquely assigned once for each input
   * instance, this unique id is then passed to all children through context
   * where it can be used to wire together aria attributes
   */
  const { current: guid } = useRef(process.env.NODE_ENV === 'test' ? 'guid' : nanoid())

  return <InputCtx.Provider value={{ guid }}>{children}</InputCtx.Provider>
}
Input.displayName = 'Input'

Input.Field = function InputField(props) {
  return elem({
    as: 'input',
    type: 'text',
    elemClassName: 'input-field',
    id: useContext(InputCtx).guid,
    ...useTheme('InputField'),
    ...props,
  })
}
Input.Field.displayName = 'InputField'

Input.Label = function InputLabel(props) {
  return elem({
    as: 'label',
    elemClassName: 'input-label',
    htmlFor: useContext(InputCtx).guid,
    ...useTheme('InputLabel'),
    ...props,
  })
}
Input.Label.displayName = 'InputLabel'

Input.Error = function InputError(props) {
  return elem({
    elemClassName: 'input-error',
    ...useTheme('InputError'),
    ...props,
  })
}
Input.Error.displayName = 'InputError'

Input.Description = function InputDescription(props) {
  return elem({
    elemClassName: 'input-description',
    ...useTheme('InputDescription'),
    ...props,
  })
}
Input.Description.displayName = 'InputDescription'

/**
 * Requirements:
 * 1. Aria attrs for label and input
 * 2. Ability to add sr-hidden to Label
 */
