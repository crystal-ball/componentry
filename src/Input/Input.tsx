/* eslint-disable @typescript-eslint/no-empty-interface */

import React, { createContext, useContext, useRef } from 'react'
import { nanoid } from 'nanoid'
import { useTheme } from '../Theme/Theme'
import { ComponentBaseProps } from '../utils/types'
import { element } from '../utils/element-creator'
import { staticComponent } from '../utils/static-component-builder'

export interface InputProps extends ComponentBaseProps<'input'> {}

export interface InputDescriptionProps extends ComponentBaseProps<'div'> {}

export interface InputErrorProps extends ComponentBaseProps<'div'> {}

export interface InputFieldProps extends ComponentBaseProps<'input'> {}

export interface InputLabelProps extends ComponentBaseProps<'label'> {}

export interface Input {
  (props: InputProps): React.ReactElement
  displayName: 'Input'
  /**
   * [Input description component üìù](https://componentry.design/components/input)
   */
  Description: React.FC<InputDescriptionProps>
  /**
   * [Input error component üìù](https://componentry.design/components/input)
   */
  Error: React.FC<InputErrorProps>
  /**
   * [Input field component üìù](https://componentry.design/components/input)
   */
  Field: React.FC<InputFieldProps>
  /**
   * [Input label component üìù](https://componentry.design/components/input)
   */
  Label: React.FC<InputLabelProps>
}

type InputCtx = { guid: string }
const InputCtx = createContext<null | InputCtx>(null)

/**
 * [Input component üìù](https://componentry.design/components/input)
 * @experimental
 */
export const Input: Input = ({ children }: InputProps) => {
  /**
   * Guid instance property will be uniquely assigned once for each input
   * instance, this unique id is then passed to all children through context
   * where it can be used to wire together aria attributes
   */
  const { current: guid } = useRef(nanoid())

  return <InputCtx.Provider value={{ guid }}>{children}</InputCtx.Provider>
}
Input.displayName = 'Input'

// --- Description subcomponent ---

Input.Description = staticComponent('InputDescription')

// --- Error subcomponent ---

Input.Error = staticComponent('InputError')

// --- Field subcomponent ---

Input.Field = function InputField(props) {
  const ctx = useContext(InputCtx)
  assertContext(ctx)

  return element('InputField', {
    as: 'input',
    type: 'text',
    id: ctx.guid, // aria -> htmlFor
    ...useTheme<Input>('InputField'),
    ...props,
  })
}
Input.Field.displayName = 'InputField'

// --- Label subcomponent ---

Input.Label = function InputLabel(props) {
  const ctx = useContext(InputCtx)
  assertContext(ctx)

  return element('InputLabel', {
    as: 'label',
    htmlFor: ctx.guid, // aria -> id
    ...useTheme('InputLabel'),
    ...props,
  })
}
Input.Label.displayName = 'InputLabel'

// --------------------------------------------------------
// Utils

/**
 * Utility asserts ctx presence for safe access
 */
function assertContext(ctx: null | InputCtx): asserts ctx is InputCtx {
  if (!ctx) throw new Error('Modal context cannot be used outside of Modal component')
}
