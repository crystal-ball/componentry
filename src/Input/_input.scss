/* componentry/src/Input/input */

.ðŸ…²-input-field {
  display: block;
  width: 100%;
  padding: $input-padding-y $input-padding-x;
  font-size: $font-size-base;
  line-height: $input-line-height;
  color: $input-color;
  background-color: $input-bg;
  background-clip: padding-box;
  border: $input-border-width solid $input-border-color;

  // Note: This has no effect on <select>s in some browsers, due to the limited stylability of `<select>`s in CSS.
  @if $enable-rounded {
    // Manually use the if/else instead of the mixin to account for iOS override
    border-radius: $input-border-radius;
  } @else {
    // Otherwise undo the iOS default
    border-radius: 0;
  }

  box-shadow: box-shadow($input-box-shadow);
  @include transition($input-transition);

  // Customize the `:focus` state to imitate native WebKit styles.
  &:focus {
    color: $input-focus-color;
    background-color: $input-focus-bg;
    border-color: $input-focus-border-color;
    outline: 0;
    // Avoid using mixin so we can pass custom focus shadow properly
    @if $enable-shadows {
      box-shadow: $input-box-shadow, $input-focus-box-shadow;
    } @else {
      box-shadow: $input-focus-box-shadow;
    }
  }

  // Placeholder
  &::placeholder {
    color: $input-placeholder-color;
    font-style: $input-placeholder-font-style;
    // Override Firefox's unusual default opacity; see https://github.com/twbs/bootstrap/pull/11526.
    opacity: 1;
  }

  // Disabled and read-only inputs
  //
  // HTML5 says that controls under a fieldset > legend:first-child won't be
  // disabled if the fieldset is disabled. Due to implementation difficulty, we
  // don't honor that edge case; we style them as disabled anyway.
  &:disabled,
  &[readonly] {
    background-color: $input-disabled-bg;
    // iOS fix for unreadable disabled content; see https://github.com/twbs/bootstrap/issues/11655.
    opacity: 1;
  }
}

// --------------------------------------------------------
// Validation

// Use :invalid with HTML and `was-validated`
// OR use .invalid

.ðŸ…²-input-field.invalid {
  background: $input-invalid-bg;
  color: $input-invalid-color;
  border-color: $input-invalid-border-color;

  // Note: This has no effect on <select>s in some browsers, due to the limited stylability of `<select>`s in CSS.
  @if $enable-rounded {
    // Manually use the if/else instead of the mixin to account for iOS override
    border-radius: $input-invalid-border-radius;
  } @else {
    // Otherwise undo the iOS default
    border-radius: 0;
  }

  &::placeholder {
    color: $input-invalid-color;
  }
}

// --------------------------------------------------------
// Sizes

.input-field-sm {
  padding: $input-padding-y-sm $input-padding-x-sm;
  font-size: $font-size-sm;
  line-height: $input-line-height-sm;
  @include border-radius($input-border-radius-sm);
}

.input-field-lg {
  padding: $input-padding-y-lg $input-padding-x-lg;
  font-size: $font-size-lg;
  line-height: $input-line-height-lg;
  @include border-radius($input-border-radius-lg);
}

// --------------------------------------------------------
// Label

.ðŸ…²-input-label {
  display: inline-block;
  margin-bottom: $input-label-margin-bottom;
  color: $input-label-color;
  font-size: $input-label-font-size;
  letter-spacing: $input-label-letter-spacing;
  text-transform: $input-label-text-transform;
}

// --------------------------------------------------------
// Description

.ðŸ…²-input-description {
  display: block;
  margin: $input-description-margin-y $input-description-margin-x;
  color: $input-description-color;
  font-size: $input-description-font-size;
  font-style: $input-description-font-style;
}

// --------------------------------------------------------
// Error state

.ðŸ…²-input-error {
  background: $input-error-bg;
  padding: $input-error-padding-y $input-error-padding-x;
  color: $input-error-color;
  font-size: $input-error-font-size;
  border-radius: $input-error-border-radius;
}
