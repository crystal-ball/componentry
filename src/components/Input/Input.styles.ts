import { getMergedConfig } from '../../plugin-postcss/configs'

const { theme } = getMergedConfig()

//                                        <Input /> styles
// --------------------------------------------------------

export const inputStyles = {
  // FIELD
  '.C9Y-InputField': {
    display: 'block',
    width: '100%',
    padding: `${theme.spacing[2]} ${theme.spacing[4]}`,
    fontSize: theme.fontSize.body,
    lineHeight: 1,
    color: theme.colors.gray[700],
    backgroundColor: theme.colors.background,
    backgroundClip: 'padding-box',
    border: theme.border.DEFAULT,
    borderRadius: theme.borderRadius.DEFAULT, // Note: This has no effect on <select>s in some browsers, due to the limited style-ability of `<select>`s in CSS.

    transition: 'borderColor 0.2s linear',

    // Customize the `:focus` state to imitate native WebKit styles.
    '&:focus': {
      borderColor: theme.colors.primary[500],
      outline: 0,
      // box-shadow: $input-box-shadow, $input-focus-box-shadow,
    },

    // Placeholder
    '&::placeholder': {
      color: theme.colors.gray[600],
      // Override Firefox's unusual default opacity, see https://github.com/twbs/bootstrap/pull/11526.
      opacity: 1,
    },

    // Disabled and read-only inputs
    //
    // HTML5 says that controls under a fieldset > legend:first-child won't be
    // disabled if the fieldset is disabled. Due to implementation difficulty, we
    // don't honor that edge case, we style them as disabled anyway.
    '&:disabled, &[readonly]': {
      backgroundColor: theme.colors.gray[200],
      // iOS fix for unreadable disabled content, see https://github.com/twbs/bootstrap/issues/11655.
      opacity: 1,
    },

    // Invalid inputs
    '&:invalid, &.C9Y-invalid': {
      background: theme.colors.error[100],
      color: theme.colors.error[500],
      borderColor: theme.colors.error[500],
    },
  },

  // LABEL
  '.C9Y-InputLabel': {
    display: 'inline-block',
    marginBottom: theme.spacing[1],
    color: theme.colors.gray[700],
    fontSize: theme.fontSize.body,
  },

  // DESCRIPTION
  '.C9Y-InputDescription': {
    display: 'block',
    margin: theme.spacing[1],
    color: theme.colors.gray[700],
    fontSize: theme.fontSize.small,
  },

  // Error
  '.C9Y-InputError': {
    display: 'block',
    margin: theme.spacing[1],
    color: theme.colors.error[500],
    fontSize: theme.fontSize.small,
  },
}

// // --------------------------------------------------------
// // Sizes

// .input-field-sm {
//   padding: $input-padding-y-sm $input-padding-x-sm,
//   font-size: $font-size-sm,
//   line-height: $input-line-height-sm,
//   @include border-radius($input-border-radius-sm),
// }

// .input-field-lg {
//   padding: $input-padding-y-lg $input-padding-x-lg,
//   font-size: $font-size-lg,
//   line-height: $input-line-height-lg,
//   @include border-radius($input-border-radius-lg),
// }
