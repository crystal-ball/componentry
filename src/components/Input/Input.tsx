import React, { createContext, useContext, useRef } from 'react'
import { nanoid } from 'nanoid'
import { useTheme } from '../Theme/Theme'
import { type ComponentBaseProps } from '../../utils/base-types'
import { element } from '../../utils/element-creator'
import { staticComponent } from '../../utils/static-component-builder'

export interface InputProps extends ComponentBaseProps<'input'> {}

export interface InputDescriptionProps extends ComponentBaseProps<'div'> {}

export interface InputErrorProps extends ComponentBaseProps<'div'> {}

export interface InputFieldProps extends ComponentBaseProps<'input'> {
  invalid?: boolean
}

export interface InputLabelProps extends ComponentBaseProps<'label'> {}

export interface Input {
  (props: InputProps): React.ReactElement
  displayName: 'Input'
  /**
   * [Input description component üìù](https://componentry.design/components/input)
   */
  Description: React.FC<InputDescriptionProps>
  /**
   * [Input error component üìù](https://componentry.design/components/input)
   */
  Error: React.FC<InputErrorProps>
  /**
   * [Input field component üìù](https://componentry.design/components/input)
   */
  Field: React.FC<InputFieldProps>
  /**
   * [Input label component üìù](https://componentry.design/components/input)
   */
  Label: React.FC<InputLabelProps>
}

type InputCtx = { guid: string }
const InputCtx = createContext<null | InputCtx>(null)

/**
 * [Input component üìù](https://componentry.design/components/input)
 * @experimental
 */
export const Input: Input = ({ children }: InputProps) => {
  /**
   * Guid instance property will be uniquely assigned once for each input
   * instance, this unique id is then passed to all children through context
   * where it can be used to wire together aria attributes
   */
  const { current: guid } = useRef(nanoid())

  return <InputCtx.Provider value={{ guid }}>{children}</InputCtx.Provider>
}
Input.displayName = 'Input'

// --- Description sub-component ---

Input.Description = staticComponent('InputDescription')

// --- Error sub-component ---

Input.Error = staticComponent('InputError')

// --- Field sub-component ---

Input.Field = function InputField(props) {
  const ctx = useContext(InputCtx)
  assertContext(ctx)

  const { invalid = false, ...rest } = {
    ...useTheme<InputFieldProps>('InputField'),
    ...props,
  }

  return element({
    as: 'input',
    type: 'text',
    id: ctx.guid, // aria -> htmlFor
    invalid: invalid ? true : undefined,
    componentCx: [
      'C9Y-InputField',
      {
        'C9Y--invalid': invalid,
      },
    ],
    ...rest,
  })
}
Input.Field.displayName = 'InputField'

// --- Label sub-component ---

Input.Label = function InputLabel(props) {
  const ctx = useContext(InputCtx)
  assertContext(ctx)

  return element({
    as: 'label',
    htmlFor: ctx.guid, // aria -> id
    componentCx: 'C9Y-InputLabel',
    ...useTheme<InputLabelProps>('InputLabel'),
    ...props,
  })
}
Input.Label.displayName = 'InputLabel'

// --------------------------------------------------------
// Utils

/**
 * Utility asserts ctx presence for safe access
 */
function assertContext(ctx: null | InputCtx): asserts ctx is InputCtx {
  if (!ctx) throw new Error('Input context cannot be used outside of Input component')
}
