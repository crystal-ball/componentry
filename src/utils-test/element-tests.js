/* eslint-disable import/no-extraneous-dependencies */
import React from 'react'
import { mount } from 'enzyme'

import dt from './dt'

/**
 * Library default component requirment test suite.
 *
 * IMPORTANT NOTES
 * - Enzyme `find()` will find **EVERY** instance of a prop, this makes it difficult
 *   to assert on find lengths, because it will be variable for components with
 *   HOCs which is why we use `toBeTruthy` to only check that some instance of the
 *   expected element attr exists
 */

export default (TestComponent, enzymeOptions = {}) => {
  /*
   * Test that any html tag or component can be passed through the 'as' prop to
   * override the parent element the component renders
   */
  test('should render as specified html element or component', () => {
    // Create a component to validate that the TestComponent returns.
    const TestAs = () => <section />
    const wrapper = mount(<TestComponent as={TestAs} />, enzymeOptions)
    expect(wrapper.find(TestAs).length).toBeTruthy()
  })

  /*
   * All components should spread passed props on the parent container for max
   * customization
   */
  test('should pass through classnames and attributes', () => {
    const wrapper = mount(
      <TestComponent className="test-custom" data-test="test-custom" />,
      enzymeOptions,
    )
    expect(wrapper.find('.test-custom').length).toBeTruthy()
    expect(wrapper.find(dt('test-custom')).length).toBeTruthy()
  })

  // TODO: should render passed children
}
