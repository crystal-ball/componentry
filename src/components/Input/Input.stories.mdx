import { Meta, Story } from '@storybook/addon-docs'

import { Input } from './Input'

<Meta title='Components/Input' component={Input} />

# Input

<Story name='Input'>
  <Input>
    <Input.Label>Componentry input</Input.Label>
    <Input.Field placeholder='type in me!' />
    <Input.Description>Really important input!</Input.Description>
  </Input>
</Story>

## Error state

<Input>
  <Input.Label>Componentry input</Input.Label>
  <Input.Field invalid required />
  <Input.Error>Oh no!</Input.Error>
</Input>

## Screen reader labels

Try to always include a label for inputs, they can be hidden using the `.sr-only` class
for screen readers.

<div className='C9Y-FormGroup-base'>
  <Input>
    <Input.Label className='sr-only'>Storybook</Input.Label>
    <Input.Field />
  </Input>
</div>
