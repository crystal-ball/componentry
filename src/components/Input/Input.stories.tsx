import type { Meta, StoryObj } from '@storybook/react'

import { Input } from './Input'

const meta: Meta<typeof Input> = {
  component: Input,
}

export default meta
type Story = StoryObj<typeof Input>

export const Primary: Story = {
  args: {
    children: (
      <>
        <Input.Label>Componentry input</Input.Label>
        <Input.Field placeholder='type in me!' />
        <Input.Description>Really important input!</Input.Description>
      </>
    ),
  },
}

export const ErrorState: Story = {
  args: {
    children: (
      <>
        <Input.Label>Componentry input</Input.Label>
        <Input.Field invalid required />
        <Input.Error>Oh no!</Input.Error>
      </>
    ),
  },
}

export const WithScreenReaderOnlyLabel: Story = {
  args: {
    children: (
      <>
        <Input.Label className='sr-only'>Storybook</Input.Label>
        <Input.Field />
      </>
    ),
  },
}
