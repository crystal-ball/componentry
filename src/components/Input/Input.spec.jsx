import { render, screen } from '@testing-library/react'
import { describe, it, expect } from 'vitest'

import { elementTests } from '../../test/element-tests'
import { Input } from './Input'

function LabelTest(testProps) {
  return (
    <Input>
      <Input.Label {...testProps} />
    </Input>
  )
}
LabelTest.displayName = Input.Label.displayName

describe('<Input />', () => {
  // Basic library element test suite
  // elementTests(Input) - no wrapper element to test...
  elementTests(LabelTest)
  // elementTests(Input.Field) - TODO: cannot test Input because it cannot have children
  elementTests(Input.Description)
  elementTests(Input.Error)
})

// Snapshots
// ---------------------------------------------------------------------------
describe('<Input /> snapshots', () => {
  it('renders correctly', () => {
    render(
      <div data-testid='input'>
        <Input>
          <Input.Label>Storybook</Input.Label>
          <Input.Field />
          <Input.Description>Really important input!</Input.Description>
          <Input.Error>Oh no!</Input.Error>
        </Input>
      </div>,
    )

    expect(screen.getByTestId('input')).toMatchSnapshot()
  })
})
