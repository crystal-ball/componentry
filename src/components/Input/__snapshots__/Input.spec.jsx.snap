// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Input /> snapshots > renders correctly 1`] = `
<div
  data-testid="input"
>
  <label
    class="C9Y-InputLabel"
    for="test-guid"
  >
    Storybook
  </label>
  <input
    class="C9Y-InputField"
    id="test-guid"
    type="text"
  />
  <div
    class="C9Y-InputDescription"
  >
    Really important input!
  </div>
  <div
    class="C9Y-InputError"
  >
    Oh no!
  </div>
</div>
`;
